v2/traversal/resumption.go

```diff
@@ -0,0 +1,257 @@
+package traversal
+
+// Resumption is an extension to an ipld traversal Progress struct that tracks the tree of the dag as it is discovered.
+// For each link, it tracks the offset that node would appear at from the beginning of the traversal, if the traversal
+// were to be serialized in a car format (e.g. [size || cid || block]*, no car header offset is included)
+// It can then resume the traversal based on either a path within the traversal, or a car offset.
+
+import (
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/ipld/go-car/v2/internal/loader"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+	"github.com/ipld/go-ipld-prime/traversal"
+)
+
+type pathNode struct {
+	link     datamodel.Link
+	offset   uint64
+	children map[datamodel.PathSegment]*pathNode
+}
+
+func newPath(link datamodel.Link, at uint64) *pathNode {
+	return &pathNode{
+		link:     link,
+		offset:   at,
+		children: make(map[datamodel.PathSegment]*pathNode),
+	}
+}
+
+func (pn pathNode) addPath(p []datamodel.PathSegment, link datamodel.Link, at uint64) {
+	if len(p) == 0 {
```

Comment:
is this just for the root case where we've made a `newPath()` and don't want to add the root to itself?